torch/optim/lr_scheduler.py

```diff
@@ -29,15 +30,32 @@ def __init__(self, optimizer, last_epoch=-1):
         # Following https://github.com/pytorch/pytorch/issues/20124
         # We would like to ensure that `lr_scheduler.step()` is called after
         # `optimizer.step()`
-        def with_counter(func, opt):
+        def with_counter(method):
+            if getattr(method, '_with_counter', False):
+                # `optimizer.step()` has already been replaced, return.
+                return method
+
+            # Keep a weak reference to the optimizer instance to prevent
+            # cyclic references.
+            instance_ref = weakref.ref(method.__self__)
+            # Get the unbound method for the same purpose.
+            func = method.__func__
+            cls = instance_ref().__class__
+            del method
+
             @wraps(func)
             def wrapper(*args, **kwargs):
-                opt._step_count += 1
-                return func(*args, **kwargs)
+                instance = instance_ref()
```

Comment:
Is it provable that the optimizer instance is always live at this point? If not maybe we should check the return result of instance...

Reply:
I think it should be, since this function will only be accessible through `optimizer.step()`.